tcmur_cmd_handler.c

```diff
@@ -181,6 +181,340 @@ struct write_same {
 	size_t iov_len;
 };
 
+struct unmap_state {
+	pthread_mutex_t lock;
+	unsigned int refcount;
+	bool error;
+};
+
+struct unmap_descriptor {
+	uint64_t offset;
+	uint64_t length;
+
+	struct tcmulib_cmd *origcmd;
+};
+
+static void handle_unmap_cbk(struct tcmu_device *dev, struct tcmulib_cmd *ucmd, int ret)
+{
+	struct unmap_descriptor *desc = ucmd->cmdstate;
+	struct tcmulib_cmd *origcmd = desc->origcmd;
+	struct unmap_state *state = origcmd->cmdstate;
+	bool error;
+
+	free(desc);
+	free(ucmd);
+
+	pthread_mutex_lock(&state->lock);
+	if (--state->refcount > 0) {
+		pthread_mutex_unlock(&state->lock);
+		return;
+	}
+	error = state->error;
```

Comment:
See PR 243. I think this will hit a issue where if the Nth unmap call failed, but the N + 1 completed ok, ret will be SAM_STAT_GOOD and the sense will not get copied over and the error will not be returned to the initiator.